src/meltano/cli/lock.py

```diff
@@ -0,0 +1,94 @@
+"""Lock command."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import click
+import structlog
+
+from meltano.core.plugin import PluginType
+from meltano.core.plugin_lock_service import (
+    LockfileAlreadyExistsError,
+    PluginLockService,
+)
+from meltano.core.project_plugins_service import DefinitionSource, ProjectPluginsService
+from meltano.core.tracking import CliContext, CliEvent, PluginsTrackingContext, Tracker
+
+from . import cli
+from .params import pass_project
+
+if TYPE_CHECKING:
+    from meltano.core.project import Project
+
+
+__all__ = ["lock"]
+logger = structlog.get_logger(__name__)
+
+
+@cli.command(short_help="Lock plugin definitions.")
+@click.argument(
+    "plugin-type",
+    type=click.Choice(PluginType.cli_arguments()),
+    required=False,
+)
+@click.argument("plugin_name", nargs=-1, required=False)
+@click.option("--update", "-u", is_flag=True, help="Update the lock file.")
+@pass_project()
+def lock(
+    project: Project,
+    plugin_type: str | None,
+    plugin_name: tuple[str, ...],
+    update: bool,
+):
+    """Lock plugin definitions.
+
+    \b\nRead more at https://docs.meltano.com/reference/command-line-interface#lock
+    """
+    tracker = Tracker(project)
+    tracker.add_contexts(
+        CliContext.from_command_and_kwargs(
+            "lock",
+            None,
+            update=update,
+            plugin_type=plugin_type,
+            plugin_name=plugin_name,
+        )
+    )
+    tracker.track_command_event(CliEvent.started)
+
+    lock_service = PluginLockService(project)
+    plugins_service = ProjectPluginsService(project)
+
+    with plugins_service.use_preferred_source(DefinitionSource.HUB):
+        if plugin_type:
+            plugin_type = PluginType.from_cli_argument(plugin_type)
+            plugins = plugins_service.get_plugins_of_type(plugin_type)
+            if plugin_name:
+                plugins = [plugin for plugin in plugins if plugin.name in plugin_name]
+        else:
+            # Make it a list so source preference is not lazily evaluated.
+            plugins = list(plugins_service.plugins())
```

Comment:
I know its ugly, but do you maybe wanna wrap 64-71 in a try except so we can fire a tracking event.(like when plugin or plugin type is not found)?

 `PluginType.from_cli_argument` and the `two plugins_service` calls can all raise exceptions directly or indirectly.

Reply:
@pandemicsyn Makes sense. Is `CliEvent.failed` the right event to push in the `except` block?

Reply:
In this case probably `CliEvent.aborted`. I've been trying to use `CliEvent.aborted` to roughly signal when the issue was user induced (i.e. they specified a non-existent plugin) and `CliEvent.failed` for a meltano induced issue. Its imperfect though and we may just end up using one failure state in the long run.

Reply:
> Its imperfect though and we may just end up using one failure state in the long run.

Related: https://github.com/meltano/meltano/issues/6041

Reply:
@pandemicsyn I've added the try/except and used `CliEvent.aborted`

@WillDaSilva #6041 looks like something we need sooner rather than later.